### PR TITLE
Refactor/gql tracker store

### PR DESCRIPTION
**Pull request by GuillaumeTech**
works with: https://github.com/botfront/botfront/pull/521/files

After Merging and releasing a new rasa, do not use it in botfront until https://github.com/botfront/botfront/pull/521 is merged because the old trackerstores are not supported

### CODE REVIEW

File: rasa_addons/core/tracker_stores/botfront.py [hunk 1]
```diff
@@ -38,134 +69,119 @@ def _start_sweeper(tracker_store, break_time):
...
-        return (
-            UnsynchronizedTracker()
-        )  # Return an object to represent synchronization failure with database
+            response = self.graphql_endpoint(query, params)
+            if "errors" in response and response["errors"]:
+                raise urllib.error.URLError("Null response.")
+            return response["data"]
+        except urllib.error.URLError:
+            logger.debug(f"something went wrong with the query {GET_TRACKER}")
+            return None
```
Comment (pheel) [inline rasa_addons/core/tracker_stores/botfront.py]: If you return None on error, then your fetch, insert and update methods can't do data.get(...), return an empty dict instead.

File: rasa_addons/core/tracker_stores/botfront.py [hunk 2]
```diff
@@ -38,134 +69,119 @@ def _start_sweeper(tracker_store, break_time):
...
-                self.trackers[sender_id] = serialized_tracker
-            else:
-                logger.debug("{} -> UNSYNCHRONIZED TRACKER UPDATE".format(sender_id))
-                self.trackers[sender_id] = UnsynchronizedTracker(serialized_tracker)
-
+        if tracker is None:  # the tracker does not exist localy ( first save)
+            updated_info = self._insert_tracker_gql(sender_id, serialized_tracker)
+            self.trackers[sender_id] = serialized_tracker
+            # update the last index and last time stamp for future uses
+            self._store_tracker_info(sender_id, updated_info)
```
Comment (pheel) [inline rasa_addons/core/tracker_stores/botfront.py]: only if update_info is not None

File: rasa_addons/core/tracker_stores/botfront.py [hunk 3]
```diff
@@ -38,134 +69,119 @@ def _start_sweeper(tracker_store, break_time):
...
             tracker_shallow_copy["events"] = new_events
-
-            logger.debug(
-                "{} -> UPDATING TRACKER: {}".format(sender_id, tracker_shallow_copy)
-            )
-            self._post(sender_id, "update", tracker_shallow_copy)
+            # only send the new events to the remote tracker
+            updated_info = self._update_tracker_gql(sender_id, tracker_shallow_copy)
+            # update the last index and last time stamp for future uses
+            self._store_tracker_info(sender_id, updated_info)
```
Comment (pheel) [inline rasa_addons/core/tracker_stores/botfront.py]: only if update_info is not None

File: rasa_addons/core/tracker_stores/botfront.py [hunk 4]
```diff
@@ -174,21 +190,48 @@ def _init_tracker(self, sender_id, tracker):
...
+            self.trackers[sender_id] = remote_tracker
+            return remote_tracker
 
+    def retrieve(self, sender_id):
+        last_index = self._get_last_index(sender_id)
+        # retreive all new info since the last sync (given by last index)
+        new_tracker_info = self._fetch_tracker(sender_id, last_index)
+        current_tracker = self.trackers.get(sender_id)
+        # the tracker exist localy and on the remote and new info have been received
+        if new_tracker_info is not None:
```
Comment (pheel) [inline rasa_addons/core/tracker_stores/botfront.py]: equivalent:

```
if not current_tracker: return None
if new_tracker_info:
    ...
return self._convert_tracker(sender_id, current_tracker)
```

File: tests/core/test_tracker_store_sync.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    
+    testTrackerStore._fetch_tracker = MagicMock(return_value = tracker2)
+    testTrackerStore.retrieve('test')
+    assert (
+            testTrackerStore.trackers['test'] == merged_tracker_1
+        ) 
+
+
+
+def test_should_not_update_events():
```
Comment (pheel) [inline tests/core/test_tracker_store_sync.py]: say why

File: rasa_addons/core/tracker_stores/botfront.py (hunk 2 above)
Comment (GuillaumeTech) [reply rasa_addons/core/tracker_stores/botfront.py]: I'have added the check in _store_tracker_info

**Comment by GuillaumeTech** [comment]
WIP : The synchronisation mechanism when using multiple rasa instances is not working fully

**Comment by GuillaumeTech** [comment]
> WIP : The synchronisation mechanism when using multiple rasa instances is not working fully

resolved

**Pull request merged**